### PR TITLE
Clean partial search

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
I didn't had time to comment David's correction of a bug.

I believe there are code duplication that is easy to remove, and simplify the logic.

The mocked test was testing internal elements that present no actual interest to the CardBrowser, so I replaced by a test which look at state we actually want to consider. It also test that the exact good number of card is loaded (thanks David for noticing the bug)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java [hunk 1]
```diff
@@ -1147,28 +1143,11 @@ public SearchCards(String query, boolean order, int numCardsToRender, int column
...
+            PartialSearch partialSearch = new PartialSearch(searchResult, mColumn1Index, mColumn2Index, mNumCardsToRender, collectionTask, col);
+            List<Long> searchResult_ = col.findCards(mQuery, mOrder, partialSearch);
             Timber.d("The search found %d cards", searchResult_.size());
-            int position = 0;
-            for (Long cid : searchResult_) {
-                CardBrowser.CardCache card = new CardBrowser.CardCache(cid, col, position++);
-                searchResult.add(card);
-            }
-            // Render the first few items
-            for (int i = 0; i < Math.min(mNumCardsToRender, searchResult.size()); i++) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:1158]: Why was this removed?

File: AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java [hunk 2]
```diff
@@ -1147,28 +1143,11 @@ public SearchCards(String query, boolean order, int numCardsToRender, int column
...
-            }
-            // Finish off the task
-            if (collectionTask.isCancelled()) {
-                Timber.d("doInBackgroundSearchCards was cancelled so return null");
-                return null;
-            } else {
-                return searchResult;
-            }
+            partialSearch.doProgress(searchResult_);
+            return searchResult;
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:1150]: What're the semantics involved with the change of return type? Does it matter

File: AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java (hunk 1 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:1158]: Because this whole code essentially duplicates what is done in PartialSearch.doProgress, so just using the doProgress method avoid code duplication

File: AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java (hunk 2 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:1150]: The return type is the same. I assume you mean the fact that we don't check a last time whether the task is cancelled. This is only because this was useless. This would only be triggered if the task is cancelled between last iteration of the loop and returning anyway, and if the task is cancelled, even if we return the list, it won't be shown. It's not like having either a list or null change anything about the action of the listener when the task is cancelled

File: AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java (hunk 1 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:1158]: Because this whole code essentially duplicates what is done in PartialSearch.doProgress, so just using the doProgress method avoid code duplication

Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:1158]: If I'm understanding the code correctly, `PartialSearch.doProgress` won't occur until a certain number of elements have been sent to it, so this will now mean that no items will be rendered if the list is less than that threshold

File: AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java (hunk 2 above)
Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:1150]: Ok

File: AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java (hunk 1 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:1158]: There is a new call to `doProgress`, added, a few line below. So it's sure that `doProgress` will be called at the end of the `task`. 
If it was called previously because there are a lot of cards, then `doProgress` will only add the new cards and won't load anything else.

If it was never called because the search returned few result, this call will fill the list and load all of those cards.

Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:1158]: Can we come up with a better design while we're here, this complexity makes me wary
